community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MuninnPageCache.java

```diff
@@ -801,6 +867,92 @@ private void clearEvictorException()
         evictorException = null;
     }
 
+    /**
+     * Scan through all the pages, flushing the dirty ones. Aim to only spend at most 50% of its time doing IO, in an
+     * effort to avoid saturating the IO subsystem or steal precious IO resources from more important work.
+     */
+    void continuouslyFlushPages()
+    {
+        Thread thread = Thread.currentThread();
+        flushThread = thread;
+
+        while ( !thread.isInterrupted() )
+        {
+            boolean flushedAnything = flushAtIORatio( backgroundFlushIoRatio );
+            if ( !flushedAnything )
+            {
+                LockSupport.parkNanos( this, TimeUnit.SECONDS.toNanos( 1 ) );
+            }
+        }
+    }
+
+    private boolean flushAtIORatio( double ratio )
+    {
+        Thread thread = Thread.currentThread();
+        long sleepDebtNanos = 0;
+        long sleepPaymentThreshold = TimeUnit.MILLISECONDS.toNanos( 10 );
+        boolean flushedAnything = false;
+
+        try ( MajorFlushEvent event = tracer.beginCacheFlush() )
+        {
+            for ( MuninnPage page : pages )
+            {
+                if ( page == null || thread.isInterrupted() )
+                {
+                    // Null pages means the page cache has been closed.
+                    thread.interrupt();
+                    return true;
+                }
+
+                // The rate is the percentage of time that we want to spend doing IO. If the rate is 0.3, then we
+                // want to spend 30% of our time doing IO. We would then spend the other 70% of the time just
+                // sleeping. This means that for every IO we do, we measure how long it takes. We can then compute
+                // the amount of time we need to sleep. Basically, if we spend 30 microseconds doing IO, then we need
+                // to sleep for 70 microseconds, with the 0.3 ratio. To get the sleep time, we can divide the IO time
+                // T by the ratio R, and then multiply the result by 1 - R. This is equivalent to (T/R) - T = S.
+                // Then, because we don't want to sleep too frequently in too small intervals, we sum up our S's and
+                // only sleep when we have collected a sleep debt of at least 10 milliseconds.
+                long stamp = page.tryReadLock();
+                if ( stamp != 0 )
+                {
+                    try
+                    {
+                        if ( !page.isDirty() )
+                        {
+                            continue; // Continue looping to the next page.
+                        }
+
+                        long startNanos = System.nanoTime();
+                        page.flush( event.flushEventOpportunity() );
```

Comment:
Yes indeed, missed that.

As another comment on top of my own comment. If the write does not include the "complete" I/O cost (queue theory) then we can undershoot the debt, and with context-switch we can overshoot.